### PR TITLE
refactor: Slight logic and docs cleanup of the clustering functions

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Overview

A general code cleanup and logic improvement for the clustering utils, namely `xgboost_distances_r2` and `hclust` functions.

Description of the changes proposed in this pull request:
* Reduce unnecessary indentation in nested for loops by using itertools `product` and `combinations_with_replacement` where appropriate
* Early exit if linkage is not valid (`xgboost_distances_r2` is potentially time-consuming)
* Add some more tests for hclust
* Some other docstring fixes and improvements (e.g. wrong dtype, add references to scipy docs, clarify return types/shapes of numpy arrays)

I'll highlight some other changes directly as code review snippets.

## Checklist

- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [ ] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

**Comment by thatlittleboy** [comment]
Note: `mypy` is complaining now because of

```
shap/plots/_decision.py:526: error: Need type annotation for "xmin"  [var-annotated]
shap/plots/_decision.py:526: error: Argument 1 has incompatible type "tuple[Any, float | ndarray[Any, Any]]"; expected "_SupportsArray[dtype[Never]] | _NestedSequence[_SupportsArray[dtype[Never]]]"  [arg-type]
shap/plots/_decision.py:527: error: Need type annotation for "xmax"  [var-annotated]
shap/plots/_decision.py:527: error: Argument 1 has incompatible type "tuple[Any, float | ndarray[Any, Any]]"; expected "_SupportsArray[dtype[Never]] | _NestedSequence[_SupportsArray[dtype[Never]]]"  [arg-type]
```
This is fixed in my other PR #3760 , so I won't duplicate the fix here. I'll wait for #3760 to be merged in first.

File: shap/utils/_clustering.py
Comment (thatlittleboy) [inline shap/utils/_clustering.py:271]: This check is irrelevant now since we already ensure the `linkage` parameter passed in is one of the 3 acceptable values above.

File: shap/utils/_clustering.py
Comment (thatlittleboy) [inline shap/utils/_clustering.py:262]: This `it.combinations_with_replacement` is identical (element-wise) with the nested for loop, it just saves us one level of indentation.

Also note here that `dist_full` is necessarily a square matrix. So `shape[1] == shape[0]`.

image

File: shap/utils/_clustering.py
Comment (thatlittleboy) [inline shap/utils/_clustering.py:268]: Note to myself (and other maintainers):

For curiosity's sake: I did try to vectorize this operation instead of using a nested for-loop, but at least on my machine, it seems to be slower (even when `num_features = 100`) so I dropped that implementation in the end.  
Also, the bulk of the time spent is on calculating the distances (`dist_full`) anyway, so it's not really worth the effort to optimize this too much.

For reference, here's the test with the attempted "vectorized" form and the original nested for loops:
```python
def test_hclust_building_triu():
    """Tests the refactoring of building a condensed upper triangular matrix within hclust,
    when using the 'xgboost_distances_r2' metric.
    """
    pytest.importorskip("xgboost")
    from shap.utils._clustering import xgboost_distances_r2
    import time

    linkage = "single"
    X = np.random.randint(1,30, size=(500, 50))
    y = np.where(X[:, 0] > 10, 1, 0)

    dist_full = xgboost_distances_r2(X, y, random_state=64)

    t0 = time.monotonic()
    # New logic (proposed)
    if linkage == "single":
        dist_new = np.minimum(dist_full, dist_full.T)
    elif linkage == "complete":
        dist_new = np.maximum(dist_full, dist_full.T)
    elif linkage == "average":
        dist_new = (dist_full + dist_full.T) / 2
    else:
        raise ValueError("Unsupported linkage type!")
    dist_new = dist_new[np.triu_indices_from(dist_new, 1)]
    t1 = time.monotonic()
    print(f">>> New {t1 - t0}")

    # Old logic (manual)
    dist = []
    for i in range(dist_full.shape[0]):
        for j in range(i + 1, dist_full.shape[1]):
            if i != j:
                if linkage == "single":
                    dist.append(min(dist_full[i, j], dist_full[j, i]))
                elif linkage == "complete":
                    dist.append(max(dist_full[i, j], dist_full[j, i]))
                elif linkage == "average":
                    dist.append((dist_full[i, j] + dist_full[j, i]) / 2)
                else:
                    raise ValueError("Unsupported linkage type!")
    dist_old = np.array(dist)
    t2 = time.monotonic()
    print(f">>> Old {t2 - t1}")

    assert (dist_new == dist_old).all()
```

And the results when running via pytest:

```
$ pytest -r P tests/utils/test_clustering.py::test_hclust_building_triu
...
_________________________________ test_hclust_building_triu _________________________________
--------------------------------------- Captured stdout call ----------------------------------------
>>> New 0.0005963749717921019
>>> Old 0.00045070797204971313
```

File: shap/utils/_clustering.py
Comment (CloseChoice) [inline shap/utils/_clustering.py:259]: NIT: I would prefer to create the matrix here right away and then fill it line by line later. 

Pro: shape of the matrix is clear as one reads the code, memory is already allocated (though this is neglectable).
Con: we would need to enumerate and another index in the loop.

File: shap/utils/_clustering.py
Comment (CloseChoice) [reply shap/utils/_clustering.py:268]: Interesting, did you test where the break even happens (I strongly assume that numpy gets way faster for larger matrices)?

**Review by CloseChoice — APPROVED** [review]
Looks good to me, thanks for cleaning this up

File: shap/utils/_clustering.py
Comment (thatlittleboy) [reply shap/utils/_clustering.py:268]: @CloseChoice Not really, at X.shape=(500,100) (i.e., 500 samples, 100 feature dimensions), the xgboost `dist_full` was already taking 40s to run on my machine (IIRC, it's a non-trivial amount of time, let's just put it that way).  
Whereas building the linkages was performing at most 100**2 operations and finishing in <0.01s.  
The numpy vectorized way is still slower in this case.

I have the same expectations as you (that the numpy version should be faster, but I think it's unnecessary to test further :x).
In most realistic cases, we have around this order of magnitude of features anyway.

File: shap/utils/_clustering.py
Comment (thatlittleboy) [reply shap/utils/_clustering.py:259]: @CloseChoice I'm assuming you mean the creation of the `dist_list`'s array here,

Agree on the memory reallocation point, and yea it probably doesn't matter here. As for the shape:  
I think the number of dimensions is clear enough from the code (i.e., ndim=1); and the length of the dist_list (k**2-k) I would argue isn't that important here since the `dist` array is input into the scipy hclust functions right after.

Hope it's alright with you that I close this

File: shap/utils/_clustering.py
Comment (connortann) [reply shap/utils/_clustering.py:262]: @thatlittleboy if I understand correctly, is this logic:

```python
for i, j in it.combinations_with_replacement(range(len(dist_full)), 2):
    if i == j:
        continue
```

equivalent to just:

```python
for i, j in it.combinations(range(len(dist_full)), 2):
```

?

Comment (thatlittleboy) [reply shap/utils/_clustering.py:262]: Thanks @connortann , nice catch. raised a separate PR to fix this. #3776

**Pull request merged**
